Orange/widgets/utils/concurrent.py

```diff
@@ -784,3 +782,167 @@ def __call__(self, *args):
         return QMetaObject.invokeMethod(
             self.obj, self.method, self.conntype, *args)
+
+
+class TaskState(QObject):
+
+    status_changed = Signal(str)
+    _p_status_changed = Signal(str)
+
+    progress_changed = Signal(float)
+    _p_progress_changed = Signal(float)
+
+    partial_result_ready = Signal(object)
+    _p_partial_result_ready = Signal(object)
+
+    def __init__(self, *args):
+        super().__init__(*args)
+        self.__future = None
+        self.watcher = FutureWatcher()
+        self.__interruption_requested = False
+        self.__progress = 0
+        # Helpers to route the signal emits via a this object's queue.
+        # This ensures 'atomic' disconnect from signals for targets/slots
+        # in the same thread. Requires that the event loop is running in this
+        # object's thread.
+        self._p_status_changed.connect(
+            self.status_changed, Qt.QueuedConnection)
+        self._p_progress_changed.connect(
+            self.progress_changed, Qt.QueuedConnection)
+        self._p_partial_result_ready.connect(
+            self.partial_result_ready, Qt.QueuedConnection)
+
+    @property
+    def future(self):
+        # type: () -> Future
+        return self.__future
+
+    def set_status(self, text):
+        self._p_status_changed.emit(text)
+
+    def set_progress_value(self, value):
+        if round(value, 1) > round(self.__progress, 1):
+            # Only emit progress when it has changed sufficiently
+            self._p_progress_changed.emit(value)
+            self.__progress = value
+
+    def set_partial_results(self, value):
+        self._p_partial_result_ready.emit(value)
+
+    def is_interruption_requested(self):
+        return self.__interruption_requested
+
+    def start(self, executor, func=None):
+        # type: (concurrent.futures.Executor, Callable[[], Any]) -> Future
+        assert self.future is None
+        assert not self.__interruption_requested
+        self.__future = executor.submit(func)
+        self.watcher.setFuture(self.future)
+        return self.future
+
+    def cancel(self):
+        assert not self.__interruption_requested
+        self.__interruption_requested = True
+        if self.future is not None:
+            rval = self.future.cancel()
+        else:
+            # not even scheduled yet
+            rval = True
+        return rval
+
+
+class InterruptRequested(BaseException):
+    pass
+
+
+class ConcurrentWidgetMixin:
+    def __init__(self):
+        self.__executor = concurrent.futures.ThreadPoolExecutor(max_workers=1)
+        self.__task = None  # type: Optional[TaskState]
+
+    @property
+    def task(self):
+        return self.__task
+
+    def _prepare_task(self, state: TaskState) -> Callable[[], Any]:
+        raise NotImplementedError
+
+    def _set_partial_results(self, result: Any) -> None:
+        raise NotImplementedError
+
+    def _set_results(self, results: Any) -> None:
```

Comment:
I think we should rename this to `_on_done` and have it raise `NotImplementedError` instead of this default implementation, which would be overridden in most cases.